### ISSUE TITLE
[PATCH 1/4] Fix docs about English.pm

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13050]
Migrated from [rt.perl.org#118593](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=118593) (status was 'resolved')

Searchable as RT118593$

**Comment by p5pRT** [comment]
### From wreis@cpan.org


The performance penalty introduced by English\.pm for regular
expressions was fixed in 5\.18\, so update docs about it \(without
needing to use "\-no\_match\_vars" special flag\) and pointing
to English\.pm docs where necessary\.


---
 pod/perlform\.pod |    2 \+\-
 pod/perlipc\.pod  |    2 \+\-
 pod/perlsec\.pod  |    2 \+\-
 pod/perlvar\.pod  |   15 \+\-\-\-\-\-\-\-\-\-\-\-\-\-\-
 4 files changed\, 4 insertions\(\+\)\, 17 deletions\(\-\)

**Comment by p5pRT** [comment]
### From wreis@cpan.org

<details><summary>0001-Fix-docs-about-English.pm.patch</summary>

```diff
diff --git a/pod/perlform.pod b/pod/perlform.pod
index 190d12e..84b6fc5 100644
--- a/pod/perlform.pod
+++ b/pod/perlform.pod
@@ -301,7 +301,7 @@ stage in the expression to single-step the debugger through):
 
 If you use the English module, you can even read the variable names:
 
-    use English '-no_match_vars';
+    use English;
     $ofh = select(OUTF);
     $FORMAT_NAME     = "My_Other_Format";
     $FORMAT_TOP_NAME = "My_Top_Format";
diff --git a/pod/perlipc.pod b/pod/perlipc.pod
index 6f07fb7..69567e9 100644
--- a/pod/perlipc.pod
+++ b/pod/perlipc.pod
@@ -546,7 +546,7 @@ write to the filehandle you opened and your kid will find it in I<his>
 STDIN.  If you open a pipe I<from> minus, you can read from the filehandle
 you opened whatever your kid writes to I<his> STDOUT.
 
-    use English qw[ -no_match_vars ];
+    use English;
     my $PRECIOUS = "/path/to/some/safe/file";
     my $sleep_count;
     my $pid;
diff --git a/pod/perlsec.pod b/pod/perlsec.pod
index 0e535af..e480cb3 100644
--- a/pod/perlsec.pod
+++ b/pod/perlsec.pod
@@ -307,7 +307,7 @@ not called with a string that the shell could expand.  This is by far the
 best way to call something that might be subjected to shell escapes: just
 never call the shell at all.  
 
-        use English '-no_match_vars';
+        use English;
         die "Can't fork: $!" unless defined($pid = open(KID, "-|"));
         if ($pid) {           # parent
             while (<KID>) {
diff --git a/pod/perlvar.pod b/pod/perlvar.pod
index 14324a5..4796004 100644
--- a/pod/perlvar.pod
+++ b/pod/perlvar.pod
@@ -61,11 +61,7 @@ Nevertheless, if you wish to use long variable names, you need only say:
 
 at the top of your program.  This aliases all the short names to the long
 names in the current package.  Some even have medium names, generally
-borrowed from B<awk>.  To avoid a performance hit, if you don't need the
-C<$PREMATCH>, C<$MATCH>, or C<$POSTMATCH> it's best to use the C<English>
-module without them:
-
-    use English '-no_match_vars';
+borrowed from B<awk>. For more info, please see L<English>.
 
 Before you continue, note the sort order for variables.  In general, we
 first list the variables in case-insensitive, almost-lexigraphical
@@ -783,15 +779,6 @@ we have not made another match:
     $1 is Mutt; $2 is Jeff
     $1 is Wallace; $2 is Grommit
 
-If you are using Perl v5.18 or earlier, note that C<use
-English> imposes a considerable performance penalty on all regular
-expression matches in a program because it uses the C<$`>, C<$&>, and
-C<$'>, regardless of whether they occur in the scope of C<use
-English>.  For that reason, saying C<use English> in libraries is
-strongly discouraged unless you import it without the match variables:
-
-    use English '-no_match_vars'
-
 The C<Devel::NYTProf> and C<Devel::FindAmpersand>
 modules can help you find uses of these
 problematic match variables in your code.

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

Applied in 6ca3c6c679258bbb20a4445b34608d144ac7090d\.  Thanks for the
patches\!  Resolving ticket\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Applied in 6ca3c6c679258bbb20a4445b34608d144ac7090d\.  Thanks for the
patches\!  Resolving ticket\.

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Jun 22\, 2013 at 07&#8203;:07&#8203;:53AM \-0700\, Wallace Reis wrote&#8203;:
> The performance penalty introduced by English\.pm for regular
> expressions was fixed in 5\.18\, so update docs about it \(without
> needing to use "\-no\_match\_vars" special flag\) and pointing
> to English\.pm docs where necessary\.

Actually that's not strictly true\.
The full fix for $\`\,$&\,$' performance issues involves the new COW
mechanism\, which was disabled by default in 5\.18 \(and enabled by default
in 5\.19\.1\)\.

5\.18\.0 has a partial fix\, where it records the presence of the three vars
separately\.

So in 5\.18\.0\, this is fast&#8203;:

&nbsp;    perl5180o \-e'$&; $a="a"x1\_000\_000; 1 while $a =~ /a/g;'

this is extremely slow&#8203;:

&nbsp;    perl5180o \-e'$\`; $a="a"x1\_000\_000; 1 while $a =~ /a/g;'

On 5\.19\.1\, they're both fast\.

Even if had been fixed in 5\.18\.0\, I think that the text in perlvars
warning that in older versions the puncuation vars are slow\, should be
kept\.


\-\- 
"There's something wrong with our bloody ships today\, Chatfield\."
&nbsp;    \-\- Admiral Beatty at the Battle of Jutland\, 31st May 1916\.

**Comment by p5pRT** [comment]
### From wreis@cpan.org

On Mon Jun 24 06&#8203;:55&#8203;:15 2013\, davem wrote&#8203;:
> Actually that's not strictly true\.
> The full fix for $\`\,$&\,$' performance issues involves the new COW
> mechanism\, which was disabled by default in 5\.18 \(and enabled by default
> in 5\.19\.1\)\.
> \[snip\]
> Even if had been fixed in 5\.18\.0\, I think that the text in perlvars
> warning that in older versions the puncuation vars are slow\, should be
> kept\.

Right\. The English\.pm documentation has already a warning about it for
older versions\, would just require an update for mentioning 5\.19 or 5\.20
instead of 5\.18\. That is simpler than having such warning all around in
other docs which makes use of English\.pm

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Jun 24\, 2013 at 07&#8203;:05&#8203;:14AM \-0700\, Wallace Reis via RT wrote&#8203;:
> On Mon Jun 24 06&#8203;:55&#8203;:15 2013\, davem wrote&#8203;:
> > Actually that's not strictly true\.
> > The full fix for $\`\,$&\,$' performance issues involves the new COW
> > mechanism\, which was disabled by default in 5\.18 \(and enabled by default
> > in 5\.19\.1\)\.
> > \[snip\]
> > Even if had been fixed in 5\.18\.0\, I think that the text in perlvars
> > warning that in older versions the puncuation vars are slow\, should be
> > kept\.
> 
> Right\. The English\.pm documentation has already a warning about it for
> older versions\, would just require an update for mentioning 5\.19 or 5\.20
> instead of 5\.18\. That is simpler than having such warning all around in
> other docs which makes use of English\.pm

I've just pushed the following 2 commits\, which hopefully resolves any
issues I had with these changes\.


commit 142a37fdb385bb222232b286abdedf9b1daaa746
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Wed Jul 24 14&#8203;:18&#8203;:22 2013 \+0100
Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
CommitDate&#8203;: Wed Jul 24 14&#8203;:42&#8203;:43 2013 \+0100

&nbsp;    English\.pm&#8203;: update perl version where perf fixed
&nbsp;    
&nbsp;    It still said that the performance of $\`\, $&\, $' was fixed in 5\.18\.
&nbsp;    Update that to 5\.20\, since COW wasn't enabled by default in 5\.18\.


Affected files \.\.\.
&nbsp;    
&nbsp;    M    lib/English\.pm

Differences \.\.\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/lib/English.pm b/lib/English.pm
index e4ee10a..6560f5f 100644
--- a/lib/English.pm
+++ b/lib/English.pm
@@ -1,6 +1,6 @@
 package English;
 
-our $VERSION = '1.07';
+our $VERSION = '1.08';
 
 require Exporter;
 @ISA = qw(Exporter);
@@ -34,9 +34,9 @@ See L<perlvar> for a complete list of these.
 
 =head1 PERFORMANCE
 
-NOTE: This was fixed in perl 5.18.  Mentioning these three variables no
+NOTE: This was fixed in perl 5.20.  Mentioning these three variables no
 longer makes a speed difference.  This section still applies if your code
-is to run on perl 5.16 or earlier.
+is to run on perl 5.18 or earlier.
 
 This module can provoke sizeable inefficiencies for regular expressions,
 due to unfortunate implementation details.  If performance matters in






```
</details>
commit 4044502721ac7b89c6d21cf1099a3a518717eeba
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Wed Jul 24 15&#8203;:20&#8203;:22 2013 \+0100
Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
CommitDate&#8203;: Wed Jul 24 15&#8203;:20&#8203;:22 2013 \+0100

&nbsp;    perlvar\.pod&#8203;: add a separate section on $& et al
&nbsp;    
&nbsp;    Add a new separate section explaining the performance issues of $\`\, $&
&nbsp;    and $'; plus descriptions of the various workarounds like @&#8203;\-\, /p and COW\,
&nbsp;    and which perl version they were each introduced in\.
&nbsp;    
&nbsp;    Then in the entries for each individual var\, strip out any commentary
&nbsp;    about performance\, and just include a link to the new performance
&nbsp;    section\.


Affected files \.\.\.
&nbsp;    
&nbsp;    M    pod/perlvar\.pod

Differences \.\.\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/pod/perlvar.pod b/pod/perlvar.pod
index a278d10..4d869f1 100644
--- a/pod/perlvar.pod
+++ b/pod/perlvar.pod
@@ -801,16 +801,51 @@ we have not made another match:
     $1 is Mutt; $2 is Jeff
     $1 is Wallace; $2 is Grommit
 
-The C<Devel::NYTProf> and C<Devel::FindAmpersand>
-modules can help you find uses of these
-problematic match variables in your code.
+=head3 Performance issues
 
-Since Perl v5.10.0, you can use the C</p> match operator flag and the
-C<${^PREMATCH}>, C<${^MATCH}>, and C<${^POSTMATCH}> variables instead
-so you only suffer the performance penalties.
+Traditionally in Perl, any use of any of the three variables  C<$`>, C<$&>
+or C<$'> (or their C<use English> equivalents) anywhere in the code, caused
+all subsequent successful pattern matches to make a copy of the matched
+string, in case the code might subsequently access one of those variables.
+This imposed a considerable performance penalty across the whole program,
+so generally the use of these variables has been discouraged.
 
-If you are using Perl v5.20.0 or higher, you do not need to worry about
-this, as the three naughty variables are no longer naughty.
+In Perl 5.6.0 the C<@-> and C<@+> dynamic arrays were introduced that
+supply the indices of successful matches. So you could for example do
+this:
+
+    $str =~ /pattern/;
+
+    print $`, $&, $'; # bad: perfomance hit
+
+    print             # good: no perfomance hit
+	substr($str, 0,     $-[0]),
+	substr($str, $-[0], $+[0]-$-[0]),
+	substr($str, $+[0]);
+
+In Perl 5.10.0 the C</p> match operator flag and the C<${^PREMATCH}>,
+C<${^MATCH}>, and C<${^POSTMATCH}> variables were introduced, that allowed
+you to suffer the penalties only on patterns marked with C</p>.
+
+In Perl 5.18.0 onwards, perl started noting the presence of each of the
+three variables separately, and only copied that part of the string
+required; so in
+
+    $`; $&; "abcdefgh" =~ /d/
+
+perl would only copy the "abcd" part of the string. That could make a big
+difference in something like
+
+    $str = 'x' x 1_000_000;
+    $&; # whoops
+    $str =~ /x/g # one char copied a million times, not a million chars
+
+In Perl 5.20.0 a new copy-on-write system was enabled by default, which
+finally fixes all performance issues with these three variables, and makes
+them safe to use anywhere.
+
+The C<Devel::NYTProf> and C<Devel::FindAmpersand> modules can help you
+find uses of these problematic match variables in your code.
 
 =over 8
 
@@ -834,12 +869,8 @@ The string matched by the last successful pattern match (not counting
 any matches hidden within a BLOCK or C<eval()> enclosed by the current
 BLOCK).
 
-In Perl v5.18 and earlier, the use of this variable
-anywhere in a program imposes a considerable
-performance penalty on all regular expression matches.  To avoid this
-penalty, you can extract the same substring by using L</@->.  Starting
-with Perl v5.10.0, you can use the C</p> match flag and the C<${^MATCH}>
-variable to do the same thing for particular match operations.
+See L</Performance issues> above for the serious performance implications
+of using this variable (even once) in your code.
 
 This variable is read-only and dynamically-scoped.
 
@@ -850,6 +881,9 @@ X<${^MATCH}>
 
 This is similar to C<$&> (C<$MATCH>) except that it does not incur the
 performance penalty associated with that variable.
+
+See L</Performance issues> above.
+
 In Perl v5.18 and earlier, it is only guaranteed
 to return a defined value when the pattern was compiled or executed with
 the C</p> modifier.  In Perl v5.20, the C</p> modifier does nothing, so
@@ -868,13 +902,8 @@ The string preceding whatever was matched by the last successful
 pattern match, not counting any matches hidden within a BLOCK or C<eval>
 enclosed by the current BLOCK.
 
-In Perl v5.18 and earlier, the use of this variable
-anywhere in a program imposes a considerable
-performance penalty on all regular expression matches.  To avoid this
-penalty, you can extract the same substring by using L</@->.  Starting
-with Perl v5.10.0, you can use the C</p> match flag and the
-C<${^PREMATCH}> variable to do the same thing for particular match
-operations.
+See L</Performance issues> above for the serious performance implications
+of using this variable (even once) in your code.
 
 This variable is read-only and dynamically-scoped.
 
@@ -885,6 +914,9 @@ X<$`> X<${^PREMATCH}>
 
 This is similar to C<$`> ($PREMATCH) except that it does not incur the
 performance penalty associated with that variable.
+
+See L</Performance issues> above.
+
 In Perl v5.18 and earlier, it is only guaranteed
 to return a defined value when the pattern was compiled or executed with
 the C</p> modifier.  In Perl v5.20, the C</p> modifier does nothing, so
@@ -907,13 +939,8 @@ enclosed by the current BLOCK).  Example:
     /def/;
     print "$`:$&:$'\n";  	# prints abc:def:ghi
 
-In Perl v5.18 and earlier, the use of this variable
-anywhere in a program imposes a considerable
-performance penalty on all regular expression matches.
-To avoid this penalty, you can extract the same substring by
-using L</@->.  Starting with Perl v5.10.0, you can use the C</p> match flag
-and the C<${^POSTMATCH}> variable to do the same thing for particular
-match operations.
+See L</Performance issues> above for the serious performance implications
+of using this variable (even once) in your code.
 
 This variable is read-only and dynamically-scoped.
 
@@ -924,6 +951,9 @@ X<${^POSTMATCH}> X<$'> X<$POSTMATCH>
 
 This is similar to C<$'> (C<$POSTMATCH>) except that it does not incur the
 performance penalty associated with that variable.
+
+See L</Performance issues> above.
+
 In Perl v5.18 and earlier, it is only guaranteed
 to return a defined value when the pattern was compiled or executed with
 the C</p> modifier.  In Perl v5.20, the C</p> modifier does nothing, so


-- 
```
</details>

Art is anything that has a label \(especially if the label is "untitled 1"\)
